contracts/bonding/BondingVotes.sol

```diff
@@ -0,0 +1,534 @@
+     * @param _account The account whose bonding state changed
+     * @param _startRound The round from which the bonding state will be active. This is normally the next round.
+     * @param _bondedAmount From {BondingManager-Delegator-bondedAmount}
+     * @param _delegateAddress From {BondingManager-Delegator-delegateAddress}
+     * @param _delegatedAmount From {BondingManager-Transcoder-delegatedAmount}
+     * @param _lastClaimRound From {BondingManager-Delegator-lastClaimRound}
+     * @param _lastRewardRound From {BondingManager-Transcoder-lastRewardRound}
+     */
+    function checkpointBondingState(
+        address _account,
+        uint256 _startRound,
+        uint256 _bondedAmount,
+        address _delegateAddress,
+        uint256 _delegatedAmount,
+        uint256 _lastClaimRound,
+        uint256 _lastRewardRound
+    ) public virtual onlyBondingManager {
+        if (_startRound != clock() + 1) {
+            revert InvalidStartRound(_startRound, clock() + 1);
+        } else if (_lastClaimRound >= _startRound) {
+            revert FutureLastClaimRound(_lastClaimRound, _startRound - 1);
+        }
+
+        BondingCheckpoint memory previous;
+        if (hasCheckpoint(_account)) {
+            previous = getBondingCheckpointAt(_account, _startRound);
+        }
+
+        BondingCheckpointsByRound storage checkpoints = bondingCheckpoints[_account];
+
+        BondingCheckpoint memory bond = BondingCheckpoint({
+            bondedAmount: _bondedAmount,
+            delegateAddress: _delegateAddress,
+            delegatedAmount: _delegatedAmount,
+            lastClaimRound: _lastClaimRound,
+            lastRewardRound: _lastRewardRound
+        });
+        checkpoints.data[_startRound] = bond;
+
+        // now store the startRound itself in the startRounds array to allow us
+        // to find it and lookup in the above mapping
+        checkpoints.startRounds.pushSorted(_startRound);
+
+        onCheckpointChanged(_account, previous, bond);
+    }
+
+    function onCheckpointChanged(
+        address _account,
+        BondingCheckpoint memory previous,
+        BondingCheckpoint memory current
+    ) internal {
+        address previousDelegate = previous.delegateAddress;
+        address newDelegate = current.delegateAddress;
+        if (previousDelegate != newDelegate) {
+            emit DelegateChanged(_account, previousDelegate, newDelegate);
+        }
+
+        bool isTranscoder = newDelegate == _account;
+        bool wasTranscoder = previousDelegate == _account;
+        if (isTranscoder) {
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, current.delegatedAmount);
+        } else if (wasTranscoder) {
+            // if the account stopped being a transcoder, we want to emit an event zeroing its "delegate votes"
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, 0);
+        }
```

Comment:
Improving this part a bit to only send these events when the `delegatedAmount` actually changed.